src/core_atmosphere/dynamics/mpas_atm_boundaries.F

```diff
@@ -169,71 +172,140 @@ subroutine mpas_atm_update_bdy_tend(clock, streamManager, block, firstCall, ierr
         call mpas_pool_get_array(lbc, 'lbc_u', u, 2)
         call mpas_pool_get_array(lbc, 'lbc_ru', ru, 2)
         call mpas_pool_get_array(lbc, 'lbc_rho_edge', rho_edge, 2)
+        call mpas_pool_get_array(lbc, 'lbc_w', w, 2)
         call mpas_pool_get_array(lbc, 'lbc_theta', theta, 2)
         call mpas_pool_get_array(lbc, 'lbc_rtheta_m', rtheta_m, 2)
         call mpas_pool_get_array(lbc, 'lbc_rho_zz', rho_zz, 2)
         call mpas_pool_get_array(lbc, 'lbc_rho', rho, 2)
         call mpas_pool_get_array(lbc, 'lbc_scalars', scalars, 2)
 
         call mpas_pool_get_array(mesh, 'cellsOnEdge', cellsOnEdge)
-        call mpas_pool_get_dimension(mesh, 'nCells', nCells)
-        call mpas_pool_get_dimension(mesh, 'nEdges', nEdges)
-        call mpas_pool_get_dimension(lbc, 'index_qv', index_qv)
+        call mpas_pool_get_dimension(mesh, 'nCells', nCells_ptr)
+        call mpas_pool_get_dimension(mesh, 'nEdges', nEdges_ptr)
+        call mpas_pool_get_dimension(mesh, 'nVertLevels', nVertLevels_ptr)
+        call mpas_pool_get_dimension(state, 'num_scalars', nScalars_ptr)
+        call mpas_pool_get_dimension(lbc, 'index_qv', index_qv_ptr)
         call mpas_pool_get_array(mesh, 'zz', zz)
 
+        MPAS_ACC_TIMER_START('mpas_atm_update_bdy_tend [ACC_data_xfer]')
+        if (.not. firstCall) then
+            call mpas_pool_get_array(lbc, 'lbc_u', lbc_tend_u, 1)
+            call mpas_pool_get_array(lbc, 'lbc_ru', lbc_tend_ru, 1)
+            call mpas_pool_get_array(lbc, 'lbc_rho_edge', lbc_tend_rho_edge, 1)
+            call mpas_pool_get_array(lbc, 'lbc_w', lbc_tend_w, 1)
+            call mpas_pool_get_array(lbc, 'lbc_theta', lbc_tend_theta, 1)
+            call mpas_pool_get_array(lbc, 'lbc_rtheta_m', lbc_tend_rtheta_m, 1)
+            call mpas_pool_get_array(lbc, 'lbc_rho_zz', lbc_tend_rho_zz, 1)
+            call mpas_pool_get_array(lbc, 'lbc_rho', lbc_tend_rho, 1)
+            call mpas_pool_get_array(lbc, 'lbc_scalars', lbc_tend_scalars, 1)
+
+            !$acc enter data copyin(lbc_tend_u, lbc_tend_ru, lbc_tend_rho_edge, lbc_tend_w, &
+            !$acc                   lbc_tend_theta, lbc_tend_rtheta_m, lbc_tend_rho_zz, &
+            !$acc                   lbc_tend_rho, lbc_tend_scalars)
+        end if
+        !$acc enter data copyin(u, w, theta, rho, scalars)
+        !$acc enter data create(ru, rho_edge, rtheta_m, rho_zz)
+        MPAS_ACC_TIMER_STOP('mpas_atm_update_bdy_tend [ACC_data_xfer]')
+
+        ! Dereference the pointers to avoid non-array pointer for OpenACC
+        nCells = nCells_ptr
+        nEdges = nEdges_ptr
+        nVertLevels = nVertLevels_ptr
+        nScalars = nScalars_ptr
+        index_qv = index_qv_ptr
+
         ! Compute lbc_rho_zz
+
+        !$acc kernels
         zz(:,nCells+1) = 1.0_RKIND          ! Avoid potential division by zero in the following line
-        rho_zz(:,:) = rho(:,:) / zz(:,:)
+        !$acc end kernels
+
+        !$acc parallel
```

Comment:
Should we add `default(present)` to this `parallel` directive (and also to others below)?

Reply:
Thanks! Done and I've also converted the `kernels` statement to a `parallel` region, as it is not a temporary loop. Combining these `parallel` regions is a little tricky, so I've left them separate as it was.